store/postgres/src/store.rs

```diff
@@ -23,18 +24,134 @@ use graph::{tokio, tokio::timer::Interval};
+/// EntityChange matching their interesting entities. When the corresponding
+/// BlockTick comes by, we pass a BlockTick to the receiver and remove
+/// the id from the set
+#[derive(Clone)]
+struct DirtyBlocks {
+    ids: Arc<RwLock<std::collections::HashSet<String>>>,
+}
+
+impl DirtyBlocks {
+    fn new() -> Self {
+        DirtyBlocks {
+            ids: Arc::new(RwLock::new(std::collections::HashSet::new())),
+        }
+    }
+
+    fn remove(&self, id: &str) -> bool {
+        self.ids.write().unwrap().remove(id)
+    }
+
+    fn insert(&self, id: String) -> bool {
+        self.ids.write().unwrap().insert(id)
+    }
+
+    fn contains(&self, id: &str) -> bool {
+        self.ids.read().unwrap().contains(id)
+    }
+}
+
+/// Subscription translates from the internal `StoreEvent` to either
+/// `EntityChange` or `BlockTick`, depending on what the subscriber is
+/// interested in. The 'other' kind of event for a subscriber is simply
+/// dropped on the floor
+///
+/// This is not a full implementation of a `futures::sink::Sink` since
+/// we don't need that right now, but could be expanded to that
+impl Subscription {
+    fn send(
+        self,
+        e: StoreEvent,
+        id: &String,
+        dirty_blocks: &DirtyBlocks,
+    ) -> impl Future<Item = (), Error = ()> {
+        match (self, e) {
+            (Subscription::Changes { sender, .. }, StoreEvent::Change(change)) => {
+                let s = sender.send(change);
+                let unit = s.map(|_| ()).map_err(|_| ());
+                Either::A(Either::A(unit))
+            }
+            (Subscription::Ticks { sender, .. }, StoreEvent::Tick(tick)) => {
+                let dirty = dirty_blocks.remove(id);
+                if dirty {
+                    let s = sender.send(tick);
+                    let unit = s.map(|_| ()).map_err(|_| ());
+                    Either::A(Either::B(unit))
+                } else {
+                    Either::B(ok(()))
+                }
+            }
+            _ => Either::B(ok(())),
+        }
+    }
+
+    fn poll_ready(&mut self) -> Poll<(), SendError<()>> {
+        match self {
+            Subscription::Changes { sender, .. } => sender.poll_ready(),
+            Subscription::Ticks { sender, .. } => sender.poll_ready(),
+        }
+    }
+
+    fn matches(&self, event: &StoreEvent, id: &String, dirty_blocks: &DirtyBlocks) -> bool {
+        match self {
+            Subscription::Changes { entities, .. } => match event {
+                StoreEvent::Change(change) => {
+                    entities.contains(&(change.subgraph_id.clone(), change.entity_type.clone()))
+                }
+                StoreEvent::Tick(_) => false,
+            },
+            Subscription::Ticks { .. } => match event {
+                StoreEvent::Change(_) => false,
+                StoreEvent::Tick(_) => dirty_blocks.contains(id),
+            },
+        }
+    }
+
+    fn is_dirty(&self, event: &StoreEvent) -> bool {
```

Comment:
Reading this without comments,I'm asking myself : is`is_dirty` the right name here? I feel `event_matches_entities` would be a more intuitive choice.